lean4-infoview-api/src/rpcSessions.ts

```diff
@@ -148,6 +148,8 @@ class RpcSessionForFile {
         try {
             const result = await this.sessions.iface.call({ method, params, sessionId, ... pos });
             this.registerRefs(result);
+            // HACK: most of our types are `T | undefined` so try to return something matching that interface
+            if (result === null) return undefined;
```

Comment:
We could, but this is the least invasive change which doesn't require all downstream code to check for `!== null` (or use truthish checks). IMO this is justified because `null` vs `undefined` is one of the many bizarre parts of JS and it is easier to stick to always using `undefined`.

Reply:
One concrete case where this pops up is with `getInteractiveGoals`.
```
{ goals: await getInteractiveGoals(...) }
-- where { goals?: InteractiveGoals }
```
If we keep `null`, then the easy fix would be to add a `?? undefined` to the use site.  But converting `null` to `undefined` seems convenient as well.

Note that we run into exactly the same issue with `Option Foo`, where the `ToJson` instance produces `null`.  For consistency it might be better to avoid any special case hacks here (since it gives the incorrect impression that `T | undefined` is the correct RPC API for `Option T`).

Reply:
Ahhh, you correctly point out that our entire RPC API is wrong. We serialize structure fields marked with the magic `?` as missing similarly to Serde's [`skip_serializing_if`](https://docs.rs/serde_with/latest/serde_with/attr.skip_serializing_none.html). In JS accessing those fields gives `undefined`. But we cannot do this for top-level values as there is only a `null` but no `undefined` in JSON. For example
```lean
structure Foo where
  a? : Option String
  b  : Option String
  deriving ToJson

/- {"b": null} -/
#eval toJson { a? := none, b := none : Foo }
```
so all the top-level `Option SomeResponse` calls really return `T | null`. In this case I guess we should go through the whole API and adjust it, and make a breaking release. Let's do this in another PR however, and temporarily keep the hack.

Reply:
Sounds good to me.